File: frontend/src/types/category.ts
export type CategoryType = "EXPENSE" | "INCOME";

export interface Category {
  id: number;
  name: string;
  user_id: number;
  workspace_id: string;
  parent_id?: number;
  type: CategoryType;
  monthly_budget?: number | null;
  children?: Category[];
  created_at?: string;
  updated_at?: string;
  created_by?: "USER" | "SYSTEM";
  mcc_code?: number;
  is_read_only?: boolean;
}

export interface CreateCategoryRequest {
  name: string;
  parent_id?: number;
  type: CategoryType;
  monthly_budget?: number | null;
}

export interface CreateCategoryResponse {
  id: number;
  name: string;
  user_id: number;
  parent_id?: number;
  type: CategoryType;
  children?: Category[];
  created_at?: string;
  updated_at?: string;
}

export interface CategoryListResponse {
  items: Category[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface CategoryFilters {
  flat?: boolean;
  page?: number;
  size?: number;
}

export interface CategoryStatisticsResponse {
  total_categories: number;
  expense_categories: number;
  income_categories: number;
  parent_categories: number;
  child_categories: number;
}

File: category_service/app/models/category.py
from sqlalchemy import Column, Integer, String, ForeignKey, Enum, DateTime, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base
import enum

class CategoryType(enum.Enum):
    EXPENSE = "EXPENSE"
    INCOME = "INCOME"

class CategoryCreatedBy(enum.Enum):
    USER = "USER"
    SYSTEM = "SYSTEM"

class Category(Base):
    __tablename__ = "categories"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, index=True)
    user_id = Column(Integer, index=True)
    parent_id = Column(Integer, ForeignKey("categories.id"), nullable=True)
    type = Column(Enum(CategoryType), nullable=False, default=CategoryType.EXPENSE)
    
    # MCC-related fields
    mcc_code = Column(Integer, ForeignKey("mcc_codes.mcc_code"), nullable=True)
    created_by = Column(Enum(CategoryCreatedBy), nullable=False, default=CategoryCreatedBy.USER)
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(DateTime, nullable=False, server_default=func.now())

    # Self-referential relationship for parent-child categories
    parent = relationship("Category", remote_side=[id], back_populates="children")
    children = relationship("Category", back_populates="parent", cascade="all, delete-orphan")